enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,480 @@
+  - fromCIDRS:
+       - 0.0.0.0/0
+       - 0::0/0
+    rules:
+    - order: 20
+      protocol: tcp
+      action: deny
+    - order: 30
+      protocol: udp
+      action: deny
+    - order: 40
+      protocol: sctp
+      action: deny
+```
+
+In case of multiple `IngressNodeFirewall` on the same nodes and interfaces if the CIDR matches or overlap the `order` field will specify which rule applied first, lower order will be excuted first, once the rule is excuted and action is taken firewall operation will be completed, all the following rules won't be checked. 
+
+- eBPF data structures and types
+
+```c
+#define DENY XDP_DROP
+#define ALLOW XDP_PASS
+#define MAX_DST_PORTS 100
+struct ruleInfo {
+    __u32 ruleId;
+    __u8 protocol;
+    union {
+        __u8 srcAddr[4];
+        __u8 srcAddr[16];
+    } srcAddru;
+    union {
+        __u8 srcMask[4];
+        __u8 srcMask[16];
+    } srcMasku;    
+     __u16 dstPort[MAX_DST_PORTS];
+    __u8 icmpType;
+    __u8 icmpCode;
+    __u8 action;
+} __packed;
+
+#define MAX_INSIDE_TARGETS   (1024)
+#define MAX_RULES_PER_TARGET (100)
+
+struct bpf_map_def SEC("maps") ingressRulesTbl = {
+    .type = BPF_MAP_TYPE_LPM_TRIE,
+    .key_size = sizeof(struct bpf_lpm_ip_trie),
+    .value_size = sizeof(struct rulesStore) + (sizeof(struct ruleInfo)*MAX_RULES_PER_TARGET),
+    .max_entries = MAX_INSIDE_TARGETS,
+    .map_flags = BPF_F_NO_PREALLOC,
+};
+
+// using Longest perfix match in case of overlapping CIDRs we need to match to the more specific CIDR.
+struct bpf_lpm_trie_ip_key {
+    __u32 prefixlen;
+    union {
+        __u8 data[4];
+        __u8 data[16];
+    } u;
+} __packed;
+
+struct rulesStore {
+    __u32 numRules;
+    struct ruleInfo rulesInfo[0];
+} __packed;
+```
+
+- eBPF statistics
+
+Implementation will use perf events map 
+```c
+struct bpf_map_def SEC("maps") rule_stats_map = {
+        .type		= BPF_MAP_TYPE_PERF_EVENT_ARRAY,
+        .key_size	= sizeof(int),
+        .value_size	= sizeof(struct ruleStatistics),
+        .max_entries	= MAX_CPUS,
+};
+
+struct ruleStatistics {
+    __u64 pkts;
+    __u64 bytes;
+} __packed;
+```
+
+the key will be the ruleId and if we need to capture the packet's header for a specific rule we just need
+to modify the flags setting to factor in how many bytes from the packet's header we need to capture for example :-
+
+```
+#define MAX_HEADER_SIZE 512
+__u64 flags = BPF_F_CURRENT_CPU;
+
+flags |= (__u64)MAX_HEADER_SIZE << 32;
+```
+### User Stories
+
+- Design ingress node Firewall CRD and APIs.
+
+- Design and Implement ingress node firewall CNO controller to manage new CRD "reconciler" . 
+  
+- Implement eBPF tabels loader and attach to the selected interfaces.
+
+- Implement implicit firewall pinholes rules to allow fail safe in case of invalid configuration (user shoudln't try to overwrite those rules).
+
+- Implement verification webhook to validate configuration and protect against invalid rules.
+  
+- Create Ingress node eBPF OCP repro.
```

Comment:
```suggestion
- Implement eBPF tables loader and attach to the selected interfaces.

- Implement implicit firewall pinholes rules to allow fail safe in case of invalid configuration (user shouldn't try to overwrite those rules).

- Implement verification webhook to validate configuration and protect against invalid rules.
  
- Create Ingress node eBPF OCP repo.
```

I think that's supposed to be "repo" as in "repository", not "repro", right?

Reply:
right